File: backend/routes/teams.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from db.connection import get_connection

router = APIRouter()

class TeamRequest(BaseModel):
    sport: str
    league: Optional[str] = None
    user: Optional[int] = None
    season: Optional[str] = None

def take_teams(db, league: Optional[str], table_league: str, table_team: str, table_broker: str):
    with db.cursor() as cursor:
        try:
            query = f"""
                SELECT {table_team}.name 
                FROM {table_team} 
                JOIN {table_broker} ON {table_team}.id = {table_broker}.id_team 
                JOIN {table_league} ON {table_broker}.id_league = {table_league}.id 
                WHERE {table_league}.name = %%s 
                ORDER BY {table_team}.name ASC
            """
            cursor.execute(query, (league,))
            return cursor.fetchall()
        except Exception:
            return []

@router.post("/teams/show-teams")
def show_teams(data: TeamRequest):
    if not data.sport:
        raise HTTPException(status_code=400, detail="Missing sport")

    db = get_connection()
    try:
        sport = data.sport
        league = data.league
        
        if sport in ["League of legends", "League of Legends"]:
            table_league = "lol_leagues"
            table_team = "lol_teams"
            table_broker = "lol_broker"
            result = take_teams(db, league, table_league, table_team, table_broker)
        elif sport == "Counter Strike":
            table_league = "cs_leagues"
            table_team = "cs_teams"
            table_broker = "cs_broker"
            result = take_teams(db, league, table_league, table_team, table_broker)
        else:
            sport_prefix = sport.lower()
            table_league = f"{sport_prefix}_leagues_season"
            table_team = f"{sport_prefix}_teams_season"
            table_broker = f"{sport_prefix}_broker_season"

            result = take_teams(db, league, table_league, table_team, table_broker)
            if not result:
                table_league = f"{sport_prefix}_league_year"
                table_team = f"{sport_prefix}_teams_year"
                table_broker = f"{sport_prefix}_broker_year"
                result = take_teams(db, league, table_league, table_team, table_broker)

        teams_names = [row["name"] for row in result]
        return {"status": "ok", "data": teams_names}
    finally:
        db.close()

@router.post("/teams/top-teams")
def top_teams(data: TeamRequest):
    if not data.sport:
        raise HTTPException(status_code=400, detail="Missing sport")

    db = get_connection()
    try:
        sport = data.sport
        user_id = data.user
        season = data.season

        if sport in ["League of legends", "League of Legends"]:
            table_team = "lol_teams"
        elif sport == "Counter Strike":
            table_team = "cs_teams"
        else:
            sport_prefix = sport.lower()
            if season == 'season':
                table_team = f"{sport_prefix}_teams_season"
            else:
                table_team = f"{sport_prefix}_teams_years"

        with db.cursor() as cursor:
            query = f"""
                SELECT {table_team}.name, (home_count+away_count) AS total_count, images.img 
                FROM {table_team} 
                JOIN images ON {table_team}.img = images.id 
                WHERE id_user = %%s 
                ORDER BY total_count DESC LIMIT 5
            """
            cursor.execute(query, (user_id,))
            result = cursor.fetchall()

        return {"status": "ok", "data": result}
    finally:
        db.close()

@router.post("/teams/home-team")
def home_team(data: TeamRequest):
    if not data.sport:
        raise HTTPException(status_code=400, detail="Missing sport")

    db = get_connection()
    try:
        sport = data.sport
        user_id = data.user
        season = data.season

        if sport in ["League of legends", "League of Legends"]:
            table_team = "lol_teams"
        elif sport == "Counter Strike":
            table_team = "cs_teams"
        else:
            sport_prefix = sport.lower()
            if season == 'season':
                table_team = f"{sport_prefix}_teams_season"
            else:
                table_team = f"{sport_prefix}_teams_years"

        with db.cursor() as cursor:
            query = f"""
                SELECT {table_team}.name, home_count, images.img 
                FROM {table_team} 
                JOIN images ON {table_team}.img = images.id 
                WHERE id_user = %%s 
                ORDER BY home_count DESC LIMIT 1
            """
            cursor.execute(query, (user_id,))
            result = cursor.fetchone()

        return {"status": "ok", "data": result}
    finally:
        db.close()

@router.post("/teams/away-team")
def away_team(data: TeamRequest):
    if not data.sport:
        raise HTTPException(status_code=400, detail="Missing sport")

    db = get_connection()
    try:
        sport = data.sport
        user_id = data.user
        season = data.season

        if sport in ["League of legends", "League of Legends"]:
            table_team = "lol_teams"
        elif sport == "Counter Strike":
            table_team = "cs_teams"
        else:
            sport_prefix = sport.lower()
            if season == 'season':
                table_team = f"{sport_prefix}_teams_season"
            else:
                table_team = f"{sport_prefix}_teams_years"

        with db.cursor() as cursor:
            query = f"""
                SELECT {table_team}.name, away_count, images.img 
                FROM {table_team} 
                JOIN images ON {table_team}.img = images.id 
                WHERE id_user = %%s 
                ORDER BY away_count DESC LIMIT 1
            """
            cursor.execute(query, (user_id,))
            result = cursor.fetchone()

        return {"status": "ok", "data": result}
    finally:
        db.close()

File: website/api/teams.php
<?php

class Teams
{
    private $db;

    public function __construct($conn)
    {
        $this->db = $conn;
    }

    public function showTeams()
    {
        $data = json_decode(file_get_contents("php://input"), true);
        $sport = $data['sport'] ?? null;
        $league = $data['league'] ?? null;

        if (!$sport) {
            http_response_code(400);
            echo json_encode(["error" => "Missing sport"]);
            return;
        }

        switch ($sport) {
            case "League of legends":
                $tableLeague = "lol_leagues";
                $tableTeam = "lol_teams";
                $tableBroker = "lol_broker";
                $result = $this->takeTeams($league, $tableLeague, $tableTeam, $tableBroker);
                break;
            case "Counter Strike":
                $tableLeague = "cs_leagues";
                $tableTeam = "cs_teams";
                $tableBroker = "cs_broker";
                $result = $this->takeTeams($league, $tableLeague, $tableTeam, $tableBroker);
                break;
            default:
                $tableLeague = strtolower($sport) . "_league_season";
                $tableTeam = strtolower($sport) . "_teams_season";
                $tableBroker = strtolower($sport) . "_broker_season";

                $result = $this->takeTeams($league, $tableLeague, $tableTeam, $tableBroker);
                if ($result && $result->num_rows === 0) {
                    $tableLeague = strtolower($sport) . "_league_year";
                    $tableTeam = strtolower($sport) . "_teams_year";
                    $tableBroker = strtolower($sport) . "_broker_year";
                    $result = $this->takeTeams($league, $tableLeague, $tableTeam, $tableBroker);
                }
        }

        $teamsNames = [];

        if ($result) {
            while ($row = $result->fetch_assoc()) {
                $teamsNames[] = $row['name'];
            }
        }

        echo json_encode(["status" => "ok", "data" => $teamsNames]);
    }

    private function takeTeams($league, $tableLeague, $tableTeam, $tableBroker)
    {
        try {
            $sql = "SELECT DISTINCT $tableTeam.name FROM $tableTeam JOIN  $tableBroker ON $tableTeam.id = $tableBroker.id_team JOIN $tableLeague ON $tableBroker.id_league = $tableLeague.id WHERE $tableLeague.name = '$league' ORDER BY $tableTeam.name ASC";
            $query = $this->db->prepare($sql);
            if (!$query)
                return false;
            $query->execute();
            $result = $query->get_result();
            return $result;
        } catch (\Exception $e) {
            http_response_code(500);
            echo json_encode(["status" => "error", "message" => $e->getMessage()]);
            exit;
        }
    }

    public function topTeams()
    {
        $data = json_decode(file_get_contents("php://input"), true);
        $sport = $data['sport'] ?? null;
        $userID = $data['user'] ?? null;
        $season = $data['season'] ?? null;

        if (!$sport) {
            http_response_code(400);
            echo json_encode(["error" => "Missing sport"]);
            return;
        }

        switch ($sport) {
            case "League of legends":
                $tableTeam = "lol_teams";
                break;
            case "Counter Strike":
                $tableTeam = "cs_teams";
                break;
            default:
                if ($season == 'season') {
                    $tableTeam = strtolower($sport) . "_teams_season";
                }
                else {
                    $tableTeam = strtolower($sport) . "_teams_year";
                }
        }
        $sql = "SELECT $tableTeam.name, (home_count+away_count) AS total_count, images.img FROM $tableTeam JOIN images ON $tableTeam.img = images.id WHERE id_user = ? ORDER BY total_count LIMIT 5";
        $query = $this->db->prepare($sql);
        $query->bind_param("i", $userID);
        $query->execute();
        $result = $query->get_result();

        $topTeams = [];

        while ($row = $result->fetch_assoc()) {
            $topTeams[] = $row;
        }

        echo json_encode(["status" => "ok", "data" => $topTeams]);
    }

    public function homeTeam()
    {
        $data = json_decode(file_get_contents("php://input"), true);
        $sport = $data['sport'] ?? null;
        $userID = $data['user'] ?? null;
        $season = $data['season'] ?? null;

        if (!$sport) {
            http_response_code(400);
            echo json_encode(["error" => "Missing sport"]);
            return;
        }

        switch ($sport) {
            case "League of legends":
                $tableTeam = "lol_teams";
                break;
            case "Counter Strike":
                $tableTeam = "cs_teams";
                break;
            default:
                if ($season == 'season') {
                    $tableTeam = strtolower($sport) . "_teams_season";
                }
                else {
                    $tableTeam = strtolower($sport) . "_teams_year";
                }
        }
        $sql = "SELECT $tableTeam.name, home_count, images.img FROM $tableTeam JOIN images ON $tableTeam.img = images.id WHERE id_user = ? ORDER BY home_count DESC LIMIT 1";
        $query = $this->db->prepare($sql);
        $query->bind_param("i", $userID);
        $query->execute();
        $result = $query->get_result()->fetch_assoc();

        echo json_encode(["status" => "ok", "data" => $result]);
    }

    public function awayTeam()
    {
        $data = json_decode(file_get_contents("php://input"), true);
        $sport = $data['sport'] ?? null;
        $userID = $data['user'] ?? null;
        $season = $data['season'] ?? null;

        if (!$sport) {
            http_response_code(400);
            echo json_encode(["error" => "Missing sport"]);
            return;
        }

        switch ($sport) {
            case "League of legends":
                $tableTeam = "lol_teams";
                break;
            case "Counter Strike":
                $tableTeam = "cs_teams";
                break;
            default:
                if ($season == 'season') {
                    $tableTeam = strtolower($sport) . "_teams_season";
                }
                else {
                    $tableTeam = strtolower($sport) . "_teams_year";
                }
        }
        $sql = "SELECT $tableTeam.name, away_count, images.img FROM $tableTeam JOIN images ON $tableTeam.img = images.id WHERE id_user = ? ORDER BY away_count DESC LIMIT 1";
        $query = $this->db->prepare($sql);
        $query->bind_param("i", $userID);
        $query->execute();
        $result = $query->get_result()->fetch_assoc();

        echo json_encode(["status" => "ok", "data" => $result]);
    }
}
